### PR TITLE
API: disallow duplicate level names

### LINKED ISSUE #18872: Problems when accessing MultiIndex level with duplicated level names
#### Code Sample, a copy-pastable example if possible

```python
In [2]: mi = pd.MultiIndex.from_product([['i'], ['ii'], ['iii']])

In [3]: mi.rename([1,5,6]).get_level_values(1) # Interpreted as label
Out[3]: Index(['i'], dtype='object', name=1)

In [4]: mi.rename([1,5,1]).get_level_values(1) # Interpreted as index
Out[4]: Index(['ii'], dtype='object', name=5)

In [5]: mi.rename(['a',5,'a']).get_level_values('a') # ValueError is OK, KeyError is not
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/home/nobackup/repo/pandas/pandas/core/indexes/multi.py in _get_level_number(self, level)
    670                 raise ValueError('The name %s occurs multiple times, use a '
--> 671                                  'level number' % level)
    672             level = self.names.index(level)

ValueError: The name a occurs multiple times, use a level number

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-5-e8a616f9610f> in <module>()
----> 1 mi.rename(['a',5,'a']).get_level_values('a')

/home/nobackup/repo/pandas/pandas/core/indexes/multi.py in get_level_values(self, level)
    975         Index(['d', 'e', 'f'], dtype='object', name='level_2')
    976         """
--> 977         level = self._get_level_number(level)
    978         values = self._get_level_values(level)
    979         return values

/home/nobackup/repo/pandas/pandas/core/indexes/multi.py in _get_level_number(self, level)
    673         except ValueError:
    674             if not isinstance(level, int):
--> 675                 raise KeyError('Level %s not found' % str(level))
    676             elif level < 0:
    677                 level += self.nlevels

KeyError: 'Level a not found'

In [6]: mi.rename([1, 'a', 'a']).get_level_values(1) # How am I going to access the second level?!
Out[6]: Index(['i'], dtype='object', name=1)

```

#### Problem description

There are different problems, but the root cause is (I think) the same:
1. the first is trivial: the ``KeyError`` in ``In [5]:`` should not appear
2. the second is that the interpretation of an integer changes when there is a duplicate name (difference between ``Out[3]`` and ``Out[4]``)
3. the third is that in a situation like ``In [6]``, I have no way whatsoever to access the second column, since it is denoted by a duplicated name _and_ its index is also the name of another column (sure, this is a perverse example, but I suspect it can bite in some cases in which users use duplicate labels _and_ pandas internal code adds integer labels)

#### Expected Output

If we were to design this from scratch, the solution would be simple: prioritize the "index" interpretation of an integer over the "label" interpretation, so that the former is always unambiguous. Is this a too strong API change?

If the answer is "no", I will be happy to implement it, possibly with a temporary warning in those cases where the behaviour will change (that is: requested label is integer _and_ is present in the names).

If the answer is "yes", I would like to at least suppress the ``KeyError`` in ``In [5]:`` and have ``In [4]:`` raise an error rather than return a result inconsistent with ``In [3]:``.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 04db779d4c93d286bb0ab87780a85d50ec490266
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-4-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.22.0.dev0+388.g04db779d4
pytest: 3.2.3
pip: 9.0.1
setuptools: 36.7.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #18872
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
+1 on the idea. Some cases to think about: pd.NaT, np.nan, anything else where x != x

**Comment by toobaz** [comment]
> anything else where x != x

I see only three options:
1. we don't do anything: the user just knows that passing ``np.nan`` as level name is stupid because ``np.nan != np.nan``
2. we build (and keep updated) an ``Index`` out of level names, and use it to access levels with the same rules with which you access items in a ``Series``
3. just like ``None``, ``np.nan`` and friends _can_ be used as level names and _can_ be duplicated but _cannot_ be used to refer to the levels (e.g. ``get_level_values(np.nan)`` will immediately raise). We describe this behavior in the docstring of ``get_level_values()`` and in all other methods which accept a level name, we state "see ``get_level_values()``")

The current PR does 1., I would be OK with 3., while I'm skeptical that 2. is worth the effort.

(If we go for 3, I would leave it to a separate PR)

File: pandas/tests/frame/test_reshape.py [hunk 1]
```diff
@@ -560,16 +560,6 @@ def test_unstack_dtypes(self):
             assert left.shape == (3, 2)
             tm.assert_frame_equal(left, right)
 
-    def test_unstack_non_unique_index_names(self):
```
Comment (jreback) [inline pandas/tests/frame/test_reshape.py]: what does this do now?

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -609,6 +600,21 @@ def test_constructor_mismatched_label_levels(self):
...
 
+    @pytest.mark.parametrize('names', [['a', 'b', 'a'], [1, 1, 2],
+                                       [1, 'a', 1]])
+    def test_duplicate_level_names(self, names):
+        # GH18872
+        pytest.raises(ValueError, pd.MultiIndex.from_product,
+                      [[0, 1]] * 3, names=names)
+
+        # With .rename()
+        mi = pd.MultiIndex.from_product([[0, 1]] * 3)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: can you check the error messages here, use ``tm.assert_raises_regex``

File: pandas/tests/series/test_alter_axes.py [hunk 3]
```diff
@@ -214,17 +214,6 @@ def test_reorder_levels(self):
         expected = Series(np.arange(6), index=e_idx)
         assert_series_equal(result, expected)
 
-        result = s.reorder_levels([0, 0, 0])
```
Comment (jreback) [inline pandas/tests/series/test_alter_axes.py]: can you add a test that raises

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt [hunk 4]
```diff
@@ -179,6 +179,7 @@ Other API Changes
 - A :class:`Series` of ``dtype=category`` constructed from an empty ``dict`` will now have categories of ``dtype=object`` rather than ``dtype=float64``, consistently with the case in which an empty list is passed (:issue:`18515`)
 - ``NaT`` division with :class:`datetime.timedelta` will now return ``NaN`` instead of raising (:issue:`17876`)
 - All-NaN levels in a ``MultiIndex`` are now assigned ``float`` rather than ``object`` dtype, promoting consistency with ``Index`` (:issue:`17929`).
+- Levels names of a ``MultiIndex`` (when not None) are now required to be unique: trying to create a ``MultiIndex`` with repeated names will raise a ``ValueError`` (:issue:`18872`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: rebase on master. can you move to 0.23 (docs were renamed), prob easiest to just check this file from master and past in new one

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_reshape.py (hunk 1 above)
Comment (toobaz) [reply pandas/tests/frame/test_reshape.py]: It raises when you try to create the ``MultiIndex`` with duplicated name.

File: pandas/tests/series/test_alter_axes.py (hunk 3 above)
Comment (toobaz) [reply pandas/tests/series/test_alter_axes.py]: It is the new ``test_set_index_duplicate_names`` (the constructor itself raises).

**Comment by toobaz** [comment]
rebased, ping

File: pandas/tests/indexes/test_multi.py (hunk 2 above)
Comment (toobaz) [reply pandas/tests/indexes/test_multi.py]: (done)

File: pandas/core/indexes/multi.py [hunk 5]
```diff
@@ -579,23 +579,22 @@ def _set_names(self, names, level=None, validate=True):
...
+            used = set()
         else:
             level = [self._get_level_number(l) for l in level]
+            used = set(self.levels[l].name
+                       for l in set(range(self.nlevels)) - set(level))
 
         # set the name
         for l, name in zip(level, names):
+            if name is not None and name in used:
+                raise ValueError("Duplicated level name: {}.".format(name))
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can you enumerate on this and report the level number where the error is as well
e.g. ``level 2, duplicate name 'foo'`` or somesuch

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. minor request on the error message. ping on green.

**Comment by toobaz** [comment]
@jreback ping

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Comment by TomAugspurger** [comment]
FYI, this is causing failures on dask due to operations like:

```python
In [1]: import dask.dataframe as dd

In [2]: import pandas as pd

In [3]:     pdf = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6, 7, 8, 9],
   ...:                         'b': [4, 5, 6, 3, 2, 1, 0, 0, 0]},
   ...:                        index=[0, 1, 3, 5, 6, 8, 9, 9, 9]).set_index("a")
   ...:
   ...:

In [4]: pdf.groupby(pdf.index).apply(lambda x: x.b)
Out[4]:
a  a
1  1    4
2  2    5
3  3    6
4  4    3
5  5    2
6  6    1
7  7    0
8  8    0
9  9    0
Name: b, dtype: int64
```

essentially, grouping by an index and doing a `.apply` that returns a `Series` whose index is the same as the original (at least the same name). This seems like a fairly common use-case.

I think we should make this a warning for and and raise later.

**Pull request merged**